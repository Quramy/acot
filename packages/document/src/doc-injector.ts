import unified from 'unified';
import markdown from 'remark-parse';
import visit from 'unist-util-visit-parents';

const IDENTIFIER = 'acot-rules';
const NEWLINE = /\r\n|[\n\r\u2028\u2029]/;
const INITIAL_LINE = -1;

export type DocInjectorConfig = {
  markdown: string;
};

export class DocInjector {
  public inject(content: string, rules: string): string {
    const tree = unified().use(markdown).parse(content);
    const rows = {
      start: INITIAL_LINE,
      end: INITIAL_LINE,
    };

    visit(tree, (node) => {
      if (node.type === 'html' && typeof node.value === 'string') {
        const value = this._parseHtmlComment(node.value);
        switch (value) {
          case `${IDENTIFIER}:start`:
            rows.start = node.position!.start.line;
            break;
          case `${IDENTIFIER}:end`:
            rows.end = node.position!.end.line;
            break;
        }
      }
    });

    if (rows.start === INITIAL_LINE && rows.end === INITIAL_LINE) {
      throw new Error('"acot-rules:{start,end}" marker does not exist.');
    }

    if (rows.start === INITIAL_LINE) {
      throw new Error('"acot-rules:end" marker does not exist.');
    }

    if (rows.end === INITIAL_LINE) {
      throw new Error('"acot-rules:start" marker does not exist.');
    }

    const lines = content.split(NEWLINE);

    lines.splice(rows.start, rows.end - rows.start - 1, rules);

    return lines.join('\n');
  }

  private _parseHtmlComment(code: string): string {
    const match = code.match(/<!--(.*)-->/);
    return match ? match[1].trim() : '';
  }
}
