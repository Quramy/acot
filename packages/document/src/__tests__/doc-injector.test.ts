import { DocInjector } from '../doc-injector';

const INJECT_CONTENT = '--- injected!! ---';

describe('DocInjector', () => {
  test.each([
    [
      'empty',
      `
# Title

> Summary


<!-- acot-rules:start -->   
   <!-- acot-rules:end -->

- Foo
  - Bar
    - Baz

<div>
  <p>HTML Paragraph.</p>
</div>
      `,
    ],
    [
      'replace',
      `
# Title

<!-- acot-rules:start -->
ORIGINAL CONTENT
<!-- acot-rules:end -->

- Foo
  - Bar
    - Baz
      `,
    ],
    [
      'comment format',
      `
<!--acot-rules:start-->
ORIGINAL CONTENT
<!--    \t acot-rules:end \t\t          -->
      `,
    ],
  ])('inject - %s', (_, input) => {
    const injector = new DocInjector();

    expect(injector.inject(input.trim(), INJECT_CONTENT)).toMatchSnapshot();
  });

  test('inject - no opening marker', () => {
    const injector = new DocInjector();
    const markdown = '<!-- acot-rules:end -->';

    expect(() => {
      injector.inject(markdown, INJECT_CONTENT);
    }).toThrow();
  });

  test('inject - no closing marker', () => {
    const injector = new DocInjector();
    const markdown = '<!-- acot-rules:start -->';

    expect(() => {
      injector.inject(markdown, INJECT_CONTENT);
    }).toThrow();
  });

  test('inject - no markers', () => {
    const injector = new DocInjector();

    expect(() => {
      injector.inject('', INJECT_CONTENT);
    }).toThrow();
  });
});
