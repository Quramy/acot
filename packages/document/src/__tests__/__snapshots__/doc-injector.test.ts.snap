// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocInjector inject - comment format 1`] = `
"<!--acot-rules:start-->
--- injected!! ---
<!--    	 acot-rules:end 		          -->"
`;

exports[`DocInjector inject - empty 1`] = `
"# Title

> Summary


<!-- acot-rules:start -->   
--- injected!! ---
   <!-- acot-rules:end -->

- Foo
  - Bar
    - Baz

<div>
  <p>HTML Paragraph.</p>
</div>"
`;

exports[`DocInjector inject - replace 1`] = `
"# Title

<!-- acot-rules:start -->
--- injected!! ---
<!-- acot-rules:end -->

- Foo
  - Bar
    - Baz"
`;
